util/contentutil/fetcher.go

```diff
@@ -0,0 +1,70 @@
+package contentutil
+
+import (
+	"context"
+	"io"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/remotes"
+	digest "github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+func FromFetcher(f remotes.Fetcher, desc ocispec.Descriptor) content.Provider {
+	return &fetchedProvider{
+		f:    f,
+		desc: desc,
+	}
+}
+
+type fetchedProvider struct {
+	f    remotes.Fetcher
+	desc ocispec.Descriptor
+}
+
+func (p *fetchedProvider) ReaderAt(ctx context.Context, dgst digest.Digest) (content.ReaderAt, error) {
+	if dgst != p.desc.Digest {
+		return nil, errors.Wrapf(errdefs.ErrNotFound, "content %v", dgst)
+	}
+
+	rc, err := p.f.Fetch(ctx, p.desc)
+	if err != nil {
+		return nil, err
+	}
+
+	return &readerAt{Reader: rc, Closer: rc, size: p.desc.Size}, nil
+}
+
+type readerAt struct {
+	io.Reader
+	io.Closer
+	size   int64
+	offset int64
+}
+
+func (r *readerAt) ReadAt(b []byte, off int64) (int, error) {
+	if ra, ok := r.Reader.(io.ReaderAt); ok {
+		return ra.ReadAt(b, off)
+	}
+
+	if r.offset != off {
+		if seeker, ok := r.Reader.(io.Seeker); ok {
+			if _, err := seeker.Seek(off, io.SeekStart); err != nil {
+				return 0, err
+			}
+			r.offset = off
+		} else {
+			return 0, errors.Errorf("unsupported offset")
+		}
+	}
+
+	n, err := r.Reader.Read(b)
+	r.offset += int64(n)
```

Comment:
Don't we need to revert the seek position here?

Reply:
Why would we? The seeker can't be detected or manually called from the client. And if a different offset is required we will seek again from the start.